src/main/java/com/commercetools/sync/shoppinglists/utils/TextLineItemUpdateActionUtils.java

```diff
@@ -0,0 +1,270 @@
+     */
+    @Nonnull
+    public static List<UpdateAction<ShoppingList>> buildTextLineItemsUpdateActions(
+        @Nonnull final ShoppingList oldShoppingList,
+        @Nonnull final ShoppingListDraft newShoppingList,
+        @Nonnull final ShoppingListSyncOptions syncOptions) {
+
+        final boolean hasOldTextLineItems = oldShoppingList.getTextLineItems() != null
+            && !oldShoppingList.getTextLineItems().isEmpty();
+        final boolean hasNewTextLineItems = newShoppingList.getTextLineItems() != null
+            && !newShoppingList.getTextLineItems().isEmpty()
+            && newShoppingList.getTextLineItems().stream().anyMatch(Objects::nonNull);
+
+        if (hasOldTextLineItems && !hasNewTextLineItems) {
+
+            return oldShoppingList.getTextLineItems()
+                                  .stream()
+                                  .map(RemoveTextLineItem::of)
+                                  .collect(toList());
+
+        } else if (!hasOldTextLineItems) {
+
+            if (!hasNewTextLineItems) {
+                return emptyList();
+            }
+
+            return newShoppingList.getTextLineItems()
+                                  .stream()
+                                  .filter(Objects::nonNull)
+                                  .map(AddTextLineItemWithAddedAt::of)
+                                  .collect(toList());
+        }
+
+        final List<TextLineItem> oldTextLineItems = oldShoppingList.getTextLineItems();
+        final List<TextLineItemDraft> newTextLineItems = newShoppingList.getTextLineItems()
+                                                                        .stream()
+                                                                        .filter(Objects::nonNull)
+                                                                        .collect(toList());
+
+        return buildUpdateActions(oldShoppingList, newShoppingList, oldTextLineItems, newTextLineItems, syncOptions);
+    }
+
+    /**
+     * The decisions in the calculating update actions are documented on the
+     * `docs/adr/0002-shopping-lists-lineitem-and-textlineitem-update-actions.md`
+     */
+    @Nonnull
+    private static List<UpdateAction<ShoppingList>> buildUpdateActions(
+        @Nonnull final ShoppingList oldShoppingList,
+        @Nonnull final ShoppingListDraft newShoppingList,
+        @Nonnull final List<TextLineItem> oldTextLineItems,
+        @Nonnull final List<TextLineItemDraft> newTextLineItems,
+        @Nonnull final ShoppingListSyncOptions syncOptions) {
+
+        final List<UpdateAction<ShoppingList>> updateActions = new ArrayList<>();
+
+        final int minSize = Math.min(oldTextLineItems.size(), newTextLineItems.size());
+        for (int i = 0; i < minSize; i++) {
+
+            final TextLineItem oldTextLineItem = oldTextLineItems.get(i);
+            final TextLineItemDraft newTextLineItem = newTextLineItems.get(i);
+
```

Comment:
Why is it not needed to check `oldTextLineItem.getName()`?
In `LineItemUpdateActionUtils`, it checks `oldLineItem.getVariant().getSku()`

Reply:
Because it's not possible with API contracts, a variant without SKU is possible.

Reply:
https://github.com/commercetools/commercetools-sync-java/pull/625/commits/011867ee660eeedcca0bd2bcb712c882ccb474ac